File: codes/semana02/jogador_encapsulado.cpp
/*
Semana 2: Encapsulamento e Validação de Dados
Exemplo: Classe Jogador com encapsulamento em C++

Conceitos abordados:
- Atributos privados
- Getters e Setters
- Validação de dados
- Sobrecarga do operador <<
- Tratamento de exceções
*/

#include <iostream>
#include <string>
#include <stdexcept>
#include <iomanip>

using namespace std;

// ============================================
// Classe Jogador com Encapsulamento
// ============================================

class Jogador {
private:
    string _nome;
    int _nivel;
    int _hp;
    int _hp_maximo;
    
    // Método privado auxiliar
    string criar_barra_hp() const {
        int percentual = (_hp * 100) / _hp_maximo;
        int barras_cheias = percentual / 10;
        int barras_vazias = 10 - barras_cheias;
        
        string barra = "[";
        for (int i = 0; i < barras_cheias; i++) barra += "█";
        for (int i = 0; i < barras_vazias; i++) barra += "░";
        barra += "] " + to_string(_hp) + "/" + to_string(_hp_maximo);
        
        return barra;
    }

public:
    // Construtor
    Jogador(string nome, int nivel = 1, int hp = 100) {
        set_nome(nome);
        set_nivel(nivel);
        _hp_maximo = hp;
        _hp = hp;
    }
    
    // ========================================
    // Getters
    // ========================================
    
    string get_nome() const {
        return _nome;
    }
    
    int get_nivel() const {
        return _nivel;
    }
    
    int get_hp() const {
        return _hp;
    }
    
    int get_hp_maximo() const {
        return _hp_maximo;
    }
    
    // ========================================
    // Setters com Validação
    // ========================================
    
    void set_nome(string nome) {
        if (nome.empty() || nome.find_first_not_of(' ') == string::npos) {
            throw invalid_argument("Nome não pode ser vazio");
        }
        _nome = nome;
    }
    
    void set_nivel(int nivel) {
        if (nivel < 1) {
            throw invalid_argument("Nível não pode ser menor que 1");
        }
        if (nivel > 100) {
            throw invalid_argument("Nível máximo é 100");
        }
        _nivel = nivel;
    }
    
    void set_hp(int hp) {
        if (hp < 0) {
            _hp = 0;
        } else if (hp > _hp_maximo) {
            _hp = _hp_maximo;
        } else {
            _hp = hp;
        }
    }
    
    void set_hp_maximo(int hp_maximo) {
        if (hp_maximo < 1) {
            throw invalid_argument("HP máximo deve ser positivo");
        }
        _hp_maximo = hp_maximo;
        if (_hp > _hp_maximo) {
            _hp = _hp_maximo;
        }
    }
    
    // ========================================
    // Métodos Públicos
    // ========================================
    
    void receber_dano(int dano) {
        if (dano < 0) {
            throw invalid_argument("Dano não pode ser negativo");
        }
        
        set_hp(_hp - dano);
        cout << _nome << " recebeu " << dano << " de dano! HP: " 
             << _hp << "/" << _hp_maximo << endl;
        
        if (_hp == 0) {
            cout << "💀 " << _nome << " foi derrotado!" << endl;
        }
    }
    
    void curar(int quantidade) {
        if (quantidade < 0) {
            throw invalid_argument("Cura não pode ser negativa");
        }
        
        int hp_anterior = _hp;
        set_hp(_hp + quantidade);
        int hp_recuperado = _hp - hp_anterior;
        
        cout << "❤️  " << _nome << " recuperou " << hp_recuperado 
             << " HP! HP: " << _hp << "/" << _hp_maximo << endl;
    }
    
    bool esta_vivo() const {
        return _hp > 0;
    }
    
    void exibir_status() const {
        cout << *this;
    }
    
    // ========================================
    // Sobrecarga do operador <<
    // ========================================
    
    friend ostream& operator<<(ostream& os, const Jogador& j) {
        string status = j.esta_vivo() ? "VIVO" : "MORTO";
        string barra_hp = j.criar_barra_hp();
        
        os << "\n╔════════════════════════════════════════╗\n";
        os << "║       STATUS DO JOGADOR                ║\n";
        os << "╠════════════════════════════════════════╣\n";
        os << "║ Nome:   " << setw(30) << left << j._nome << " ║\n";
        os << "║ Nível:  " << setw(30) << left << j._nivel << " ║\n";
        os << "║ Status: " << setw(30) << left << status << " ║\n";
        os << "║ HP:     " << barra_hp << "\n";
        os << "╚════════════════════════════════════════╝\n";
        
        return os;
    }
};

// ============================================
// Programa de Teste
// ============================================

int main() {
    cout << "============================================================" << endl;
    cout << "SEMANA 2: Encapsulamento e Validação de Dados" << endl;
    cout << "============================================================" << endl;
    cout << endl;
    
    try {
        // Criando jogador
        Jogador jogador("Thorin", 10, 200);
        jogador.exibir_status();
        
        // Testando validações
        cout << "\n>>> Testando validações:\n" << endl;
        
        try {
            jogador.set_nivel(-5);
        } catch (const invalid_argument& e) {
            cout << "❌ Erro capturado: " << e.what() << endl;
        }
        
        try {
            jogador.set_nome("");
        } catch (const invalid_argument& e) {
            cout << "❌ Erro capturado: " << e.what() << endl;
        }
        
        try {
            jogador.receber_dano(-10);
        } catch (const invalid_argument& e) {
            cout << "❌ Erro capturado: " << e.what() << endl;
        }
        
        // Testando HP com validação automática
        cout << "\n>>> Testando controle de HP:\n" << endl;
        jogador.receber_dano(50);
        jogador.curar(30);
        jogador.curar(200);  // Tenta curar além do máximo
        
        jogador.exibir_status();
        
        // Testando dano fatal
        cout << "\n>>> Testando dano fatal:\n" << endl;
        jogador.receber_dano(300);
        jogador.exibir_status();
        
        // Tentando curar jogador morto
        cout << "\n>>> Tentando curar jogador morto:\n" << endl;
        jogador.curar(50);
        jogador.exibir_status();
        
    } catch (const exception& e) {
        cerr << "Erro fatal: " << e.what() << endl;
        return 1;
    }
    
    return 0;
}

/*
Para compilar e executar:
g++ -std=c++17 -o jogador_encapsulado jogador_encapsulado.cpp
./jogador_encapsulado
*/

File: codes/semana02/jogador_encapsulado.py
"""
Semana 2: Encapsulamento e Validação de Dados
Exemplo: Classe Jogador com encapsulamento em Python

Conceitos abordados:
- Atributos privados (convenção _)
- Propriedades (@property)
- Getters e Setters
- Validação de dados
- Método __str__
"""


class Jogador:
    """Representa um jogador no sistema de RPG com encapsulamento"""
    
    def __init__(self, nome, nivel=1, hp=100):
        """Inicializa um novo jogador"""
        self._nome = nome
        self._nivel = nivel
        self._hp_maximo = hp
        self._hp = hp
    
    # ========================================
    # Propriedades (Getters e Setters)
    # ========================================
    
    @property
    def nome(self):
        """Retorna o nome do jogador"""
        return self._nome
    
    @nome.setter
    def nome(self, valor):
        """Define o nome do jogador com validação"""
        if not valor or len(valor.strip()) == 0:
            raise ValueError("Nome não pode ser vazio")
        self._nome = valor.strip()
    
    @property
    def nivel(self):
        """Retorna o nível do jogador"""
        return self._nivel
    
    @nivel.setter
    def nivel(self, valor):
        """Define o nível com validação"""
        if valor < 1:
            raise ValueError("Nível não pode ser menor que 1")
        if valor > 100:
            raise ValueError("Nível máximo é 100")
        self._nivel = valor
    
    @property
    def hp(self):
        """Retorna HP atual"""
        return self._hp
    
    @hp.setter
    def hp(self, valor):
        """Define HP com validação (não pode exceder máximo)"""
        if valor < 0:
            self._hp = 0
        elif valor > self._hp_maximo:
            self._hp = self._hp_maximo
        else:
            self._hp = valor
    
    @property
    def hp_maximo(self):
        """Retorna HP máximo"""
        return self._hp_maximo
    
    @hp_maximo.setter
    def hp_maximo(self, valor):
        """Define HP máximo com validação"""
        if valor < 1:
            raise ValueError("HP máximo deve ser positivo")
        self._hp_maximo = valor
        # Ajusta HP atual se necessário
        if self._hp > self._hp_maximo:
            self._hp = self._hp_maximo
    
    # ========================================
    # Métodos Públicos
    # ========================================
    
    def receber_dano(self, dano):
        """
        Aplica dano ao jogador com validação
        
        Args:
            dano: Quantidade de dano (deve ser positivo)
        """
        if dano < 0:
            raise ValueError("Dano não pode ser negativo")
        
        self.hp -= dano  # Usa o setter para validação automática
        print(f"{self.nome} recebeu {dano} de dano! HP: {self.hp}/{self.hp_maximo}")
        
        if self.hp == 0:
            print(f"💀 {self.nome} foi derrotado!")
    
    def curar(self, quantidade):
        """
        Cura o jogador com validação
        
        Args:
            quantidade: Quantidade de HP a recuperar (deve ser positivo)
        """
        if quantidade < 0:
            raise ValueError("Cura não pode ser negativa")
        
        hp_anterior = self.hp
        self.hp += quantidade  # Usa o setter para validação automática
        hp_recuperado = self.hp - hp_anterior
        
        print(f"❤️  {self.nome} recuperou {hp_recuperado} HP! HP: {self.hp}/{self.hp_maximo}")
    
    def esta_vivo(self):
        """Verifica se o jogador está vivo"""
        return self.hp > 0
    
    def exibir_status(self):
        """Exibe o status do jogador"""
        print(self)
    
    # ========================================
    # Métodos Especiais
    # ========================================
    
    def __str__(self):
        """Representação em string do jogador"""
        status = "VIVO" if self.esta_vivo() else "MORTO"
        barra_hp = self._criar_barra_hp()
        
        return f"""
╔════════════════════════════════════════╗
║       STATUS DO JOGADOR                ║
╠════════════════════════════════════════╣
║ Nome:   {self.nome:<30} ║
║ Nível:  {self.nivel:<30} ║
║ Status: {status:<30} ║
║ HP:     {barra_hp}
╚════════════════════════════════════════╝
"""
    
    def _criar_barra_hp(self):
        """Cria uma barra visual de HP (método privado)"""
        percentual = (self.hp / self.hp_maximo) * 100
        barras_cheias = int(percentual / 10)
        barras_vazias = 10 - barras_cheias
        
        barra = "█" * barras_cheias + "░" * barras_vazias
        return f"[{barra}] {self.hp}/{self.hp_maximo}"


# ============================================
# Programa de Teste
# ============================================

def main():
    print("=" * 60)
    print("SEMANA 2: Encapsulamento e Validação de Dados")
    print("=" * 60)
    print()
    
    # Criando jogador
    jogador = Jogador("Thorin", nivel=10, hp=200)
    jogador.exibir_status()
    
    # Testando validações
    print("\n>>> Testando validações:\n")
    
    try:
        jogador.nivel = -5  # Deve gerar erro
    except ValueError as e:
        print(f"❌ Erro capturado: {e}")
    
    try:
        jogador.nome = ""  # Deve gerar erro
    except ValueError as e:
        print(f"❌ Erro capturado: {e}")
    
    try:
        jogador.receber_dano(-10)  # Deve gerar erro
    except ValueError as e:
        print(f"❌ Erro capturado: {e}")
    
    # Testando HP com validação automática
    print("\n>>> Testando controle de HP:\n")
    jogador.receber_dano(50)
    jogador.curar(30)
    jogador.curar(200)  # Tenta curar além do máximo (é limitado automaticamente)
    
    jogador.exibir_status()
    
    # Testando dano fatal
    print("\n>>> Testando dano fatal:\n")
    jogador.receber_dano(300)
    jogador.exibir_status()
    
    # Tentando curar jogador morto
    print("\n>>> Tentando curar jogador morto:\n")
    jogador.curar(50)
    jogador.exibir_status()


if __name__ == "__main__":
    main()
